_plugins/pa11y.rb

```diff
@@ -0,0 +1,177 @@
+  )
+
+  if document.to_scan?
+    File.open(PA11Y_TARGET_FILE, 'a') { |f|
+      f.write(file.destination(NO_ROOT_PATH) + "\n")
+    }
+  end
+end
+
+# Outputs posts and pages to scan to a file.
+Jekyll::Hooks.register :documents, :post_render do |doc|
+  check_file_for_changes(doc)
+end
+
+# The :documents hook above doesn't seem to touch the pages/ directory
+Jekyll::Hooks.register :pages, :post_render do |doc|
+  check_file_for_changes(doc)
+end
+
+# Produces a sample set of pages needed for a site-wide pa11y test.
+class SiteSampler
+
+  attr_reader :config, :site_files, :permalinks
+
+  def initialize(config, site_files=nil, permalinks=nil)
+    @config = config
+    @site_files = site_files || default_site_files
+    @permalinks = permalinks || default_permalinks
+  end
+
+  def pages
+    folders.flat_map { |folder| sample folder }
+  end
+
+  private
+
+  # Samples 3 files from a given folder, and its index
+  # @todo Blog folder should sample 3 blog posts as well _site/blog/\d{4}/**/**
+  # @todo Site folder should sample 3 site pages _site/**/index.html
+  # @todo The matcher should change, not the list of files being returned
+  def sample(folder)
+    if ["_site/", "_site/blog/"].include?(folder)
+      [ File.join(folder, "index.html") ]
+    else
+      folder_regex = Regexp.new("^" + folder)
+      index_regex = Regexp.new("^" + File.join(folder, "index.html") + "$")
+      site_files.select do |file|
+        file.match?(folder_regex) || file.match?(index_regex)
+      end.reject {|filename| filename.match?(/.pdf$/i)}.sample(3)
+    end
+  end
+
+  def folders
+    permalinks.map do |permalink|
+      File.join("_site", parameterless_path(permalink), "/")
+    end.uniq.sort
+  end
+
+  # Default list of all the files in the site
+  def default_site_files
+    Dir.glob("_site/**/*").reject { |f| File.directory?(f) }
+  end
+
+  # Default list of all the permalink URL templates as given
+  # in the config document. These form the basis for the sampled
+  # pages later on.
+  def default_permalinks
+    [
+      config["jekyll-archives"]["permalinks"].values,
+      config["collections"].map { |_name, data| data["permalink"] },
+      config["paginate_path"],
+      "blog/"
+    ].flatten.uniq
+  end
+
+  # Follows the file paths until reaching a :param
+  def parameterless_path(path)
+    Pathname.new(path).each_filename.take_while { |x| !x.match? "^:" }.join("/")
+  end
+end
+
+
+# Changes to files in these directories trigger a global check
+# @todo Is there a way to get these more dynamically?
+SITEWIDE_FOLDERS = ["assets", "_includes", "_sass"]
+
+# If files have changed in any of the folders that trigger a site-wide scan,
+# sample folders across the site and add them to the list of files to check
+# with pa11y for a given Pull Request.
+# @todo Is there a way to unnest this code? Jekyll hooks don't allow
+#   early returns.
+Jekyll::Hooks.register :site, :post_render do |site|
+  # https://rubular.com/r/nfFL9P69KHSBoY
```

Comment:
clarification: What's the context for this comment? And/or can it be removed?

Reply:
The comment is providing an example of the generated RegEx against an imagined list of changed files to highlight what the regular expression is doing in case someone needs to change the code who is unfamiliar with regex.

Reply:
Yep - I will add that to the list of changes, thanks!